core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java

```diff
@@ -88,14 +94,19 @@ public void stop() {
    public CacheTopology join(String cacheName, CacheJoinInfo joinInfo, CacheTopologyHandler stm,
          PartitionHandlingManager phm) throws Exception {
       log.debugf("Node %s joining cache %s", transport.getAddress(), cacheName);
-      LocalCacheStatus cacheStatus = new LocalCacheStatus(joinInfo, stm, phm);
+
+      // For Total Order caches, we must not move the topology updates to another thread
+      ExecutorService topologyUpdatesExecutor = joinInfo.isTotalOrder() ? new WithinThreadExecutor() : stateTransferExecutor;
```

Comment:
maybe I'm a little nitpick here but you can have a single WithinThreadExecutor shared